--- chromium-81.0.4044.92.orig/sandbox/features.gni
+++ chromium-81.0.4044.92/sandbox/features.gni
@@ -11,6 +11,7 @@ import("//build/config/nacl/config.gni")
 use_seccomp_bpf = (is_linux || is_android) &&
                   (current_cpu == "x86" || current_cpu == "x64" ||
                    current_cpu == "arm" || current_cpu == "arm64" ||
-                   current_cpu == "mipsel" || current_cpu == "mips64el")
+                   current_cpu == "mipsel" || current_cpu == "mips64el" ||
+                   current_cpu == "ppc64")
 
 use_seccomp_bpf = use_seccomp_bpf || is_nacl_nonsfi
