diff --git a/third_party/node/node.py b/third_party/node/node.py
index 8097e2c49acf..67a96fa0b03f 100755
--- a/third_party/node/node.py
+++ b/third_party/node/node.py
@@ -7,14 +7,15 @@ from os import path as os_path
 import platform
 import subprocess
 import sys
-
+from distutils.spawn import find_executable
 
 def GetBinaryPath():
-  return os_path.join(os_path.dirname(__file__), *{
-    'Darwin': ('mac', 'node-darwin-x64', 'bin', 'node'),
-    'Linux': ('linux', 'node-linux-x64', 'bin', 'node'),
-    'Windows': ('win', 'node.exe'),
-  }[platform.system()])
+  return find_executable('node')
+  #return os_path.join(os_path.dirname(__file__), *{
+  #  'Darwin': ('mac', 'node-darwin-x64', 'bin', 'node'),
+  #  'Linux': ('linux', 'node-linux-x64', 'bin', 'node'),
+  #  'Windows': ('win', 'node.exe'),
+  #}[platform.system()])
 
 
 def RunNode(cmd_parts, stdout=None):
